### PR TITLE
Calculate incounts from outcount in dependency rules

### PR DESCRIPTION
**Pull request by juliuskunze**

### CODE REVIEW

File: fastar/core.py [hunk 1]
```diff
@@ -230,3 +233,13 @@ def get_aval(x):
 
 def abstractify_lazy(invals):
   return [abstractify(x._aval) if isinstance(x, LazyArray) else x for x in invals]
+
+class Ones(jc.ShapedArray):
```
Comment (j-towns) [inline fastar/core.py]: Is it necessary to inherit from ShapedArray here? Surely the only attribute of Ones we should really be using is the shape (i.e. we'd never need to inspect its dtype).

File: fastar/core.py [hunk 2]
```diff
@@ -230,3 +233,13 @@ def get_aval(x):
...
+  def __init__(self, shape):
+    super().__init__(shape, int)
+
+def is_ones(count: Union[Array, Ones]):
+  return type(count) == Ones
```
Comment (j-towns) [inline fastar/core.py]: I could be wrong but I think
```python
type(count) is Ones
```
is more standard for this kind of check.

File: fastar/rules.py [hunk 3]
```diff
@@ -160,38 +164,39 @@ def concatenate_dependency_rule(outbox, *operands, dimension):
...
-def slice_dependency_rule(outbox, operand, start_indices, limit_indices, strides):
-  outstart, outshape = outbox
-  strides = np.ones_like(outshape) if strides is None else np.array(strides)
-  zeros = np.zeros_like(outshape)
+def slice_dependency_rule(outstart, outcount, operand, start_indices, limit_indices, strides):
+  strides = np.ones_like(outcount.shape) if strides is None else np.array(strides)
+  zeros = np.zeros_like(outcount.shape)
   return ([outstart * strides + start_indices],
-          [lax.pad(np.ones(outshape, int), 0, zip(zeros, zeros, strides - 1))],
+          [lax.pad(materialize(outcount), 0, zip(zeros, zeros, strides - 1))],
```
Comment (j-towns) [inline fastar/rules.py]: Should this be `lax_ref.pad`?

File: fastar/rules.py [hunk 4]
```diff
@@ -160,38 +164,39 @@ def concatenate_dependency_rule(outbox, *operands, dimension):
...
 dependency_rules[lax.slice_p] = slice_dependency_rule
 
-def transpose_dependency_rule(outbox, operand, permutation):
-  outstart, outshape = outbox
+def transpose_dependency_rule(outstart, outcount, operand, permutation):
   inverse_perm = np.argsort(permutation)
   return ([np.take(outstart, inverse_perm)],
-          [np.ones(np.take(outshape, inverse_perm), int)],
+          [Ones(np.take(outcount.shape, inverse_perm))
+           if is_ones(outcount) else laxref.transpose(outcount, inverse_perm)],
```
Comment (j-towns) [inline fastar/rules.py]: Could use `np.transpose` instead of `lax_ref.transpose` since IIRC the two are identical.

File: fastar/core.py (hunk 1 above)
Comment (juliuskunze) [reply fastar/core.py]: I think it's good to have things like `ndim`, `size` built-in, wdyt?

File: fastar/rules.py (hunk 3 above)
Comment (juliuskunze) [reply fastar/rules.py]: Yes.

File: fastar/core.py (hunk 1 above)
Comment (j-towns) [reply fastar/core.py]: We can easily implement those ourselves here, to me it feels slightly dangerous to inherit from ShapedArray.

Comment (j-towns) [reply fastar/core.py]: It's sort of suggesting to the reader that there might be some deep reason that we're inheriting from ShapedArray, which isn't the case.

Comment (juliuskunze) [reply fastar/core.py]: Done.

**Pull request merged**